[WIP][FrameworkBundle] Add debug autoconfigure command

aaa2000 opened the pull request:
Close #26295

| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes <!-- don't forget to update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- don't forget to update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | #26295   <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

<!--
Write a short README entry for your feature/bugfix here (replace this comment block.)
This will help people understand your PR and can be used as a start of the Doc PR.
Additionally:
 - Bug fixes must be submitted against the lowest branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the master branch.
-->

Replace the pull request #26745.

Creation of a new command as proposed in the comment https://github.com/symfony/symfony/pull/26745#issuecomment-385232838t to show the full template of the autoconfigure of interfaces/classes

<details>
 <summary>Output of the command based on modified symfony-demo application to have binding autoconfiguration, multiple tags attributes and multiple method calls </summary>

```

sf debug:autoconfiguration 

Autoconfiguration
=================


Autoconfiguration for "App\Twig\AppExtension"
==============================================

 ---------- -------------------- 
  Option     Value               
 ---------- -------------------- 
  Tag        app.twig.extension  
  Bindings   $locales: fr|en|es  
                                 
 ---------- -------------------- 

Autoconfiguration for "Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepositoryInterface"
==============================================

 -------- ----------------------------- 
  Option   Value                        
 -------- ----------------------------- 
  Tag      doctrine.repository_service  
 -------- ----------------------------- 

Autoconfiguration for "Doctrine\Bundle\FixturesBundle\ORMFixtureInterface"
==============================================

 -------- ---------------------- 
  Option   Value                 
 -------- ---------------------- 
  Tag      doctrine.fixture.orm  
 -------- ---------------------- 

Autoconfiguration for "Psr\Log\LoggerAwareInterface"
==============================================

 ------------- ---------------------------- 
  Option        Value                       
 ------------- ---------------------------- 
  Method call   - [setLogger, ['@logger']]  
                - [setLogger, ['@logger']]  
                                            
 ------------- ---------------------------- 

Autoconfiguration for "Sensio\Bundle\FrameworkExtraBundle\Request\ParamConverter\ParamConverterInterface"
==============================================

 -------- ------------------------- 
  Option   Value                    
 -------- ------------------------- 
  Tag      request.param_converter  
 -------- ------------------------- 

Autoconfiguration for "Symfony\Bridge\Monolog\Processor\ProcessorInterface"
==============================================

 -------- ------------------- 
  Option   Value              
 -------- ------------------- 
  Tag      monolog.processor  
 -------- ------------------- 

Autoconfiguration for "Symfony\Bundle\FrameworkBundle\Controller\AbstractController"
==============================================

 -------- ------------------------------ 
  Option   Value                         
 -------- ------------------------------ 
  Tag      controller.service_arguments  
 -------- ------------------------------ 

Autoconfiguration for "Symfony\Bundle\FrameworkBundle\Controller\Controller"
==============================================

 -------- ------------------------------ 
  Option   Value                         
 -------- ------------------------------ 
  Tag      controller.service_arguments  
 -------- ------------------------------ 

Autoconfiguration for "Symfony\Component\Config\ResourceCheckerInterface"
==============================================

 -------- ------------------------------- 
  Option   Value                          
 -------- ------------------------------- 
  Tag      config_cache.resource_checker  
 -------- ------------------------------- 

Autoconfiguration for "Symfony\Component\Console\Command\Command"
==============================================

 -------- ----------------- 
  Option   Value            
 -------- ----------------- 
  Tag      console.command  
 -------- ----------------- 

Autoconfiguration for "Symfony\Component\DependencyInjection\EnvVarProcessorInterface"
==============================================

 -------- ----------------------------- 
  Option   Value                        
 -------- ----------------------------- 
  Tag      container.env_var_processor  
 -------- ----------------------------- 

Autoconfiguration for "Symfony\Component\DependencyInjection\ServiceLocator"
==============================================

 -------- --------------------------- 
  Option   Value                      
 -------- --------------------------- 
  Tag      container.service_locator  
 -------- --------------------------- 

Autoconfiguration for "Symfony\Component\EventDispatcher\EventSubscriberInterface"
==============================================

 -------- ------------------------- 
  Option   Value                    
 -------- ------------------------- 
  Tag      kernel.event_subscriber  
 -------- ------------------------- 

Autoconfiguration for "Symfony\Component\Form\FormTypeGuesserInterface"
==============================================

 -------- ------------------- 
  Option   Value              
 -------- ------------------- 
  Tag      form.type_guesser  
 -------- ------------------- 

Autoconfiguration for "Symfony\Component\Form\FormTypeInterface"
==============================================

 -------- ----------- 
  Option   Value      
 -------- ----------- 
  Tag      form.type  
 -------- ----------- 

Autoconfiguration for "Symfony\Component\HttpKernel\CacheClearer\CacheClearerInterface"
==============================================

 -------- ---------------------- 
  Option   Value                 
 -------- ---------------------- 
  Tag      kernel.cache_clearer  
 -------- ---------------------- 

Autoconfiguration for "Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerInterface"
==============================================

 -------- --------------------- 
  Option   Value                
 -------- --------------------- 
  Tag      kernel.cache_warmer  
 -------- --------------------- 

Autoconfiguration for "Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface"
==============================================

 -------- ------------------------------------ 
  Option   Value                               
 -------- ------------------------------------ 
  Tag      controller.argument_value_resolver  
 -------- ------------------------------------ 

Autoconfiguration for "Symfony\Component\HttpKernel\DataCollector\DataCollectorInterface"
==============================================

 -------- ---------------- 
  Option   Value           
 -------- ---------------- 
  Tag      data_collector  
 -------- ---------------- 

Autoconfiguration for "Symfony\Component\Messenger\Handler\MessageHandlerInterface"
==============================================

 -------- --------------------------- 
  Option   Value                      
 -------- --------------------------- 
  Tag      messenger.message_handler  
 -------- --------------------------- 

Autoconfiguration for "Symfony\Component\Messenger\Transport\TransportFactoryInterface"
==============================================

 -------- ----------------------------- 
  Option   Value                        
 -------- ----------------------------- 
  Tag      messenger.transport_factory  
 -------- ----------------------------- 

Autoconfiguration for "Symfony\Component\PropertyInfo\PropertyAccessExtractorInterface"
==============================================

 -------- -------------------------------- 
  Option   Value                           
 -------- -------------------------------- 
  Tag      property_info.access_extractor  
 -------- -------------------------------- 

Autoconfiguration for "Symfony\Component\PropertyInfo\PropertyDescriptionExtractorInterface"
==============================================

 -------- ------------------------------------- 
  Option   Value                                
 -------- ------------------------------------- 
  Tag      property_info.description_extractor  
 -------- ------------------------------------- 

Autoconfiguration for "Symfony\Component\PropertyInfo\PropertyInitializableExtractorInterface"
==============================================

 -------- --------------------------------------- 
  Option   Value                                  
 -------- --------------------------------------- 
  Tag      property_info.initializable_extractor  
 -------- --------------------------------------- 

Autoconfiguration for "Symfony\Component\PropertyInfo\PropertyListExtractorInterface"
==============================================

 -------- ------------------------------ 
  Option   Value                         
 -------- ------------------------------ 
  Tag      property_info.list_extractor  
 -------- ------------------------------ 

Autoconfiguration for "Symfony\Component\PropertyInfo\PropertyTypeExtractorInterface"
==============================================

 -------- ------------------------------ 
  Option   Value                         
 -------- ------------------------------ 
  Tag      property_info.type_extractor  
 -------- ------------------------------ 

Autoconfiguration for "Symfony\Component\Security\Core\Authorization\Voter\VoterInterface"
==============================================

 -------- ---------------- 
  Option   Value           
 -------- ---------------- 
  Tag      security.voter  
 -------- ---------------- 

Autoconfiguration for "Symfony\Component\Serializer\Encoder\DecoderInterface"
==============================================

 -------- -------------------- 
  Option   Value               
 -------- -------------------- 
  Tag      serializer.encoder  
 -------- -------------------- 

Autoconfiguration for "Symfony\Component\Serializer\Encoder\EncoderInterface"
==============================================

 -------- -------------------- 
  Option   Value               
 -------- -------------------- 
  Tag      serializer.encoder  
 -------- -------------------- 

Autoconfiguration for "Symfony\Component\Serializer\Normalizer\DenormalizerInterface"
==============================================

 -------- ----------------------- 
  Option   Value                  
 -------- ----------------------- 
  Tag      serializer.normalizer  
 -------- ----------------------- 

Autoconfiguration for "Symfony\Component\Serializer\Normalizer\NormalizerInterface"
==============================================

 -------- ----------------------- 
  Option   Value                  
 -------- ----------------------- 
  Tag      serializer.normalizer  
 -------- ----------------------- 

Autoconfiguration for "Symfony\Component\Validator\ConstraintValidatorInterface"
==============================================

 -------- -------------------------------- 
  Option   Value                           
 -------- -------------------------------- 
  Tag      validator.constraint_validator  
 -------- -------------------------------- 

Autoconfiguration for "Symfony\Component\Validator\ObjectInitializerInterface"
==============================================

 -------- ----------------------- 
  Option   Value                  
 -------- ----------------------- 
  Tag      validator.initializer  
 -------- ----------------------- 

Autoconfiguration for "Symfony\Contracts\Service\ResetInterface"
==============================================

 --------------- -------------------------- 
  Option          Value                     
 --------------- -------------------------- 
  Tag             kernel.reset              
  Tag attribute   [                         
                    [                       
                      "method" => "reset"   
                    ]                       
                  ]                         
                                            
  Tag             kernel.reset2             
  Tag attribute   [                         
                    [                       
                      "method" => "reset2"  
                    ]                       
                  ]                         
                                            
 --------------- -------------------------- 

Autoconfiguration for "Symfony\Contracts\Service\ServiceSubscriberInterface"
==============================================

 -------- ------------------------------ 
  Option   Value                         
 -------- ------------------------------ 
  Tag      container.service_subscriber  
 -------- ------------------------------ 

Autoconfiguration for "Twig\Extension\ExtensionInterface"
==============================================

 -------- ---------------- 
  Option   Value           
 -------- ---------------- 
  Tag      twig.extension  
 -------- ---------------- 

Autoconfiguration for "Twig\Extension\RuntimeExtensionInterface"
==============================================

 -------- -------------- 
  Option   Value         
 -------- -------------- 
  Tag      twig.runtime  
 -------- -------------- 

Autoconfiguration for "Twig\Loader\LoaderInterface"
==============================================

 -------- ------------- 
  Option   Value        
 -------- ------------- 
  Tag      twig.loader  
 -------- ------------- 

Autoconfiguration for "Twig_ExtensionInterface"
==============================================

 -------- ---------------- 
  Option   Value           
 -------- ---------------- 
  Tag      twig.extension  
 -------- ---------------- 

Autoconfiguration for "Twig_LoaderInterface"
==============================================

 -------- ------------- 
  Option   Value        
 -------- ------------- 
  Tag      twig.loader  
 -------- ------------- 
```

</details>

aaa2000 committed:
Add debug autoconfigure command

Close #26295

aaa2000 committed:
Fix coding standard

nicolas-grekas reviewed (commented):
Thanks for working on this, here is a round of comments.

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php:
```diff
@@ -0,0 +1,153 @@
+        /** @var ChildDefinition $autoconfiguredInstanceofItem */
+        foreach ($autoconfiguredInstanceofItems as $key => $autoconfiguredInstanceofItem) {
+            $tags = $autoconfiguredInstanceofItem->getTags();
+            $tableRows = array();
```
maybe we can get rid of the `$tableRows` array by using `addRow()` and `render()`, as this is for `master`. IMO its more readable afterwards

docs: https://symfony.com/doc/current/components/console/helpers/table.html#modifying-rendered-tables

aaa2000 replied:
I'm not sure it's more readable. To have the same style that the others commands, it is necessary to configure the table.

```
diff --git a/src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php b/src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php
index bffdd32751..6d80649634 100644
--- a/src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php
+++ b/src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php
@@ -12,6 +12,7 @@
 namespace Symfony\Bundle\FrameworkBundle\Command;
 
 use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Helper\Table;
 use Symfony\Component\Console\Input\InputArgument;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Input\InputOption;
Stash this hunk [y,n,q,a,d,j,J,g,/,e,?]? y
@@ -91,26 +92,30 @@ EOF
 
         /** @var ChildDefinition $autoconfiguredInstanceofItem */
         foreach ($autoconfiguredInstanceofItems as $key => $autoconfiguredInstanceofItem) {
-            $tableRows = array();
+            $section = $output->section();
+            $table = new Table($section);
+            $table->setStyle(Table::getStyleDefinition('symfony-style-guide'));
+            $table->addRow(array('Option', 'Value'));
 
             foreach ($autoconfiguredInstanceofItem->getTags() as $tag => $tagAttributes) {
-                $tableRows[] = array('Tag', $tag);
+                $table->addRow(array('Tag', $tag));
                 if ($tagAttributes !== array(array())) {
-                    $tableRows[] = array('Tag attribute', $this->dumpTagAttribute($tagAttributes));
+                    $table->addRow(array('Tag attribute', $this->dumpTagAttribute($tagAttributes)));
                 }
             }
 
             if ($autoconfiguredInstanceofItem->getMethodCalls()) {
-                $tableRows[] = array('Method call', $this->dumpMethodCall($autoconfiguredInstanceofItem));
+                $table->addRow(array('Method call', $this->dumpMethodCall($autoconfiguredInstanceofItem)));
             }
 
             if ($autoconfiguredInstanceofItem->getBindings()) {
-                $tableRows[] = array('Bindings', $this->dumpBindings($autoconfiguredInstanceofItem));
+                $table->addRow(array('Bindings', $this->dumpBindings($autoconfiguredInstanceofItem)));
             }
 
             $io->writeln(sprintf("Autoconfiguration for \"%s\"\n==============================================", $key));
             $io->newLine();
-            $io->table(array('Option', 'Value'), $tableRows);
+            $table->render();
+            $io->newLine();
         }
     }

```

OskarStark commented:
Could you please post a screenshot of the output?

aaa2000 committed:
Fixes

Fix many setMethodCalls display
Fix autoconfiguration display
Add bindings display
Add tests with class fixtures

aaa2000 committed:
Fix coding standard

aaa2000 committed:
Remove unnecessary config

aaa2000 commented on src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php:
```diff
@@ -0,0 +1,153 @@
+            }
+            $tableRows[] = array('Public', $autoconfiguredInstanceofItem->isPublic() ? 'yes' : 'no');
+            $tableRows[] = array('Shared', $autoconfiguredInstanceofItem->isShared() ? 'yes' : 'no');
+            $tableRows[] = array('Autowired', $autoconfiguredInstanceofItem->isAutowired() ? 'yes' : 'no');
```
I removed Public, Shared and Autowired and I add bindings

aaa2000 commented on src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php:
```diff
@@ -0,0 +1,153 @@
+            }
+
+            if ($tags && array_values($tags) !== array(array(array()))) {
+                $tableRows[] = array('Tags attributes', $this->dumpTagsAttributes($tags));
```
If there are many attributes, the command displays like that 

```
 --------------- -------------------------- 
  Option          Value                     
 --------------- -------------------------- 
  Tag             kernel.reset              
  Tag attribute   [                         
                    [                       
                      "method" => "reset"   
                    ]                       
                  ]                         
                                            
  Tag             kernel.reset2             
  Tag attribute   [                         
                    [                       
                      "method" => "reset2"  
                    ]                       
                  ]                         
                                            
 --------------- -------------------------- 
```

aaa2000 commented:
@OskarStark Have you seen the output of command in the PR header, it requires to unfold. Do you prefer a screenshot ?

aaa2000 commented on src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/DebugAutoconfigurationBundle/DebugAutoconfigurationBundle.php:
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Symfony\Bundle\FrameworkBundle\Tests\Functional\Bundle\DebugAutoconfigurationBundle;
```
Maybe, I can use the TestBundle instead of create a new bundle ?

aaa2000 committed:
Fix windows tests

OskarStark commented:
Ah i missed it 😊

aaa2000 committed:
Fix display tag attributes

aaa2000 committed:
Modify expected output of tests

ro0NL reviewed (commented):
not sure we need a new command, feels like it should be part of `debug:container --autoconfiguration` instead. Re-building and YAML dumping the container here looks suspicious to me.

Alternatively extend from `ContainerDebugCommand` like `DebugAutowiringCommand` also does.

aaa2000 commented:
> Re-building and YAML dumping the container here looks suspicious to me.

Ok, I don't like it too but I didn't see other way to display method calls and bindings.
